src/etwlogger_driver.c

```diff
@@ -0,0 +1,262 @@
+}
+/*the function will also attempt to produce some human readable strings for GetLastError*/
+void etwlogger_log_with_GetLastError(const char* file, const char* func, int line, const char* format, ...)
+{
+    DWORD lastError;
+    char* lastErrorAsString;
+
+
+    va_list args;
+    va_start(args, format);
+
+    /*this is what this case will do:
+    1. snip the last error
+    2. create a string with what that last error means
+    3. printf the system message (__FILE__, __LINE__ etc) + the last error + whatever the user wanted
+    */
+    /*1. snip the last error*/
+    lastError = GetLastError();
+
+    /*2. create a string with what that last error means*/
+    SYSTEMTIME t;
+    GetSystemTime(&t);
+
+    lastErrorAsString = lastErrorToString(lastError);
+    if (lastErrorAsString == NULL)
+    {
+        char* userMessage = vprintf_alloc(format, args);
+        if (userMessage == NULL)
+        {
+            if (EventWriteLogErrorEvent("unable to print user error or last error", file, &t, func, line) != ERROR_SUCCESS)
+            {
+                (void)printf("failure in EventWriteLogErrorEvent");
+            }
+        }
+        else
+        {
+            if (EventWriteLogErrorEvent(userMessage, file, &t, func, line) != ERROR_SUCCESS)
+            {
+                (void)printf("failure in EventWriteLogErrorEvent");
+            }
+            free(userMessage);
+        }
+    }
+    else
+    {
+        char* userMessage = vprintf_alloc(format, args);
+        if (userMessage == NULL)
+        {
+            if (EventWriteLogErrorEvent(lastErrorAsString, file, &t, func, line) != ERROR_SUCCESS)
+            {
+                (void)printf("failure in EventWriteLogErrorEvent");
+            }
+        }
+        else
+        {
+            if (EventWriteLogLastError(userMessage, file, &t, func, line, lastErrorAsString) != ERROR_SUCCESS)
+            {
+                (void)printf("failure in EventWriteLogErrorEvent");
+            }
+            free(userMessage);
+        }
+        free(lastErrorAsString);
+    }
+
+    
+    va_end(args);
+}
+
+
+static int isETWLoggerInit = 0;
+
+void etwlogger_log(LOG_CATEGORY log_category, const char* file, const char* func, int line, unsigned int options, const char* format, ...)
+{
+    (void)options;
+    /*lazily init the logger*/
+    if (isETWLoggerInit == 0)
+    {
+        if (EventRegisterMicrosoft_ServiceBus() != ERROR_SUCCESS)
+        {
+            /*go back to printf, maybe somebody notices*/
+            printf("failure in EventRegisterMicrosoft_ServiceBus");
+        }
+        else
+        {
+            /*should check if the provider is registered with the system... maybe at a later time*/
+            isETWLoggerInit = 1; /*and stays 1 until the process exits*/ /*sorry, no graceful exit with EventUnregisterMicrosoft_ServiceBus*/
+            LogInfo("EventRegisterMicrosoft_ServiceBus success"); /*selflogging that the log service has started*/
+        }
+    }
+    
```

Comment:
Awesome catch! I have put it in a static function.

---
In reply to: [126215732](https://github.com/Azure/azure-c-shared-utility/pull/89#discussion_r126215732) [](ancestors = 126215732)